File: android/app/src/main/java/com/iunico/mynotes/NotificationSlots.java
package com.iunico.mynotes;

import java.util.Calendar;

/**
 * NotificationSlots
 *
 * Helper Java que replica la lógica de notification-slots.ts para la capa Android nativa.
 * Asigna a cada ítem (evento, hábito, tarea) una franja horaria determinística
 * usando un hash djb2 de su ID, distribuiendo las notificaciones a lo largo del día
 * en lugar de dispararlas todas al mismo tiempo.
 *
 * Franjas predeterminadas: 09:00, 13:00, 18:00 (espejo de DEFAULT_SLOTS en TS).
 */
public final class NotificationSlots {

    // ─────────────────────────────────────────────────────────────────────────
    // Configuración de franjas
    // ─────────────────────────────────────────────────────────────────────────

    /** Horas de las franjas (índice compartido con SLOT_MINUTES). */
    private static final int[] SLOT_HOURS   = { 9, 13, 18 };
    /** Minutos de las franjas. */
    private static final int[] SLOT_MINUTES = { 0,  0,  0 };

    // ─────────────────────────────────────────────────────────────────────────
    // Hash determinístico (djb2) — mismo algoritmo que notification-slots.ts
    // ─────────────────────────────────────────────────────────────────────────

    /**
     * Convierte un String en un entero no negativo de 32 bits usando djb2.
     * Produce el mismo resultado que {@code hashStringToInt} en notification-slots.ts.
     */
    public static int hashStringToInt(String str) {
        if (str == null || str.isEmpty()) return 0;
        int hash = 5381;
        for (int i = 0; i < str.length(); i++) {
            // hash * 33 + c  (equivalente a (hash << 5) + hash + c)
            hash = ((hash << 5) + hash) + str.charAt(i);
        }
        return Math.abs(hash);
    }

    // ─────────────────────────────────────────────────────────────────────────
    // Asignación de franja
    // ─────────────────────────────────────────────────────────────────────────

    /**
     * Devuelve el índice de franja (0-based) para el itemId dado.
     * Determinístico: el mismo ID siempre produce el mismo índice.
     */
    public static int assignSlotIndex(String itemId) {
        return hashStringToInt(itemId) % SLOT_HOURS.length;
    }

    /**
     * Devuelve la hora de la franja asignada al itemId (0-23).
     */
    public static int assignSlotHour(String itemId) {
        return SLOT_HOURS[assignSlotIndex(itemId)];
    }

    /**
     * Devuelve el minuto de la franja asignada al itemId (0-59).
     */
    public static int assignSlotMinute(String itemId) {
        return SLOT_MINUTES[assignSlotIndex(itemId)];
    }

    // ─────────────────────────────────────────────────────────────────────────
    // Construcción del triggerAtMillis
    // ─────────────────────────────────────────────────────────────────────────

    /**
     * Calcula el {@code triggerAtMillis} para programar la alarma del ítem.
     *
     * <ul>
     *   <li>Si la franja aún no ha pasado hoy → dispara hoy en esa franja.</li>
     *   <li>Si la franja ya pasó hoy → dispara mañana en esa franja.</li>
     *   <li>Se aplica un jitter aleatorio de ±{@code jitterMinutes} para que
     *       ítems del mismo slot no lleguen en el mismo segundo exacto.</li>
     * </ul>
     *
     * @param itemId        ID del ítem (evento, tarea, hábito).
     * @param jitterMinutes Máximo desfase aleatorio en minutos (usar 5 normalmente).
     * @return              Timestamp en milisegundos para pasar a AlarmManager.
     */
    public static long buildTriggerMillis(String itemId, int jitterMinutes) {
        int slotHour   = assignSlotHour(itemId);
        int slotMinute = assignSlotMinute(itemId);

        Calendar cal = Calendar.getInstance();
        cal.set(Calendar.HOUR_OF_DAY, slotHour);
        cal.set(Calendar.MINUTE, slotMinute);
        cal.set(Calendar.SECOND, 0);
        cal.set(Calendar.MILLISECOND, 0);

        // Aplicar jitter ±jitterMinutes
        long jitterMs = 0;
        if (jitterMinutes > 0) {
            // Rango [-jitterMinutes, +jitterMinutes] en milisegundos
            jitterMs = (long) ((Math.random() * 2 - 1) * jitterMinutes * 60 * 1000L);
            cal.setTimeInMillis(cal.getTimeInMillis() + jitterMs);
        }

        // Si la hora calculada ya pasó, mover al mismo slot de mañana
        if (cal.getTimeInMillis() <= System.currentTimeMillis()) {
            cal.add(Calendar.DAY_OF_YEAR, 1);
            // Re-establecer hora/minuto para no arrastrar el jitter al día siguiente
            cal.set(Calendar.HOUR_OF_DAY, slotHour);
            cal.set(Calendar.MINUTE, slotMinute);
            cal.set(Calendar.SECOND, 0);
            cal.set(Calendar.MILLISECOND, 0);
            if (jitterMinutes > 0) {
                jitterMs = (long) ((Math.random() * 2 - 1) * jitterMinutes * 60 * 1000L);
                cal.setTimeInMillis(cal.getTimeInMillis() + jitterMs);
            }
        }

        return cal.getTimeInMillis();
    }

    // Constructor privado — clase de utilidad estática
    private NotificationSlots() {}
}

File: src/lib/notification-slots.ts
/**
 * notification-slots.ts
 *
 * Utilidad central para distribuir notificaciones en franjas horarias fijas.
 *
 * Problema que resuelve: cuando todos los hábitos/tareas se programan para el
 * mismo momento, se disparan en ráfaga. Esta utilidad asigna a cada ítem una
 * franja determinística (basada en el hash de su ID) para que los recordatorios
 * lleguen repartidos a lo largo del día.
 *
 * Las franjas son persistibles en @capacitor/preferences para que en el futuro
 * el usuario pueda personalizarlas desde ajustes.
 */

import { Preferences } from '@capacitor/preferences';

// ─────────────────────────────────────────────────
// Tipos
// ─────────────────────────────────────────────────

export interface TimeSlot {
    hour: number;   // 0-23
    minute: number; // 0-59
}

// ─────────────────────────────────────────────────
// Configuración por defecto
// ─────────────────────────────────────────────────

/**
 * Franjas horarias predeterminadas: mañana, mediodía y tarde.
 * Editables por el usuario en el futuro (ver saveNotificationSlots).
 */
export const DEFAULT_SLOTS: TimeSlot[] = [
    { hour: 9,  minute: 0 },
    { hour: 13, minute: 0 },
    { hour: 18, minute: 0 },
];

const PREFS_KEY = 'notification-slots';

// ─────────────────────────────────────────────────
// Hash determinístico (djb2)
// ─────────────────────────────────────────────────

/**
 * Convierte un string en un entero de 32 bits usando el algoritmo djb2.
 * Siempre devuelve un número no negativo.
 *
 * El mismo ID siempre produce el mismo hash, garantizando que el ítem
 * quede en la misma franja aunque se reprograme o edite.
 */
export function hashStringToInt(str: string): number {
    let hash = 5381;
    for (let i = 0; i < str.length; i++) {
        // hash * 33 + charCode  (equivalente a (hash << 5) + hash + c)
        hash = ((hash << 5) + hash) + str.charCodeAt(i);
        hash |= 0; // fuerza entero de 32 bits con signo
    }
    return Math.abs(hash);
}

// ─────────────────────────────────────────────────
// Asignación de franjas
// ─────────────────────────────────────────────────

/**
 * Devuelve el índice de franja (0-based) para el itemId dado.
 * Es determinístico: el mismo ID siempre produce el mismo índice.
 */
export function assignSlotIndex(itemId: string, slots: TimeSlot[] = DEFAULT_SLOTS): number {
    return hashStringToInt(itemId) % slots.length;
}

/**
 * Devuelve la franja `TimeSlot` asignada a itemId dentro del array de slots.
 */
export function assignSlot(itemId: string, slots: TimeSlot[] = DEFAULT_SLOTS): TimeSlot {
    return slots[assignSlotIndex(itemId, slots)];
}

// ─────────────────────────────────────────────────
// Construcción de la fecha de disparo
// ─────────────────────────────────────────────────

/**
 * Construye un Date con la hora/minuto de la franja asignada, basado en `baseDate`.
 *
 * - Si la hora calculada ya pasó hoy, se devuelve para mañana.
 * - `jitterMinutes` añade un desplazamiento aleatorio de ±jitterMinutes para
 *   que ítems del mismo slot no disparen en el mismo segundo exacto.
 *
 * @param baseDate     Fecha de referencia (normalmente `new Date()`).
 * @param slot         Franja horaria asignada.
 * @param jitterMinutes Máximo desfase aleatorio en minutos (por defecto 5).
 */
export function buildTriggerDate(
    baseDate: Date,
    slot: TimeSlot,
    jitterMinutes = 5,
): Date {
    const trigger = new Date(baseDate);
    trigger.setHours(slot.hour, slot.minute, 0, 0);

    // Aplicar jitter ±jitterMinutes
    if (jitterMinutes > 0) {
        const jitterMs = (Math.random() * 2 - 1) * jitterMinutes * 60 * 1000;
        trigger.setTime(trigger.getTime() + jitterMs);
    }

    // Si la hora ya pasó hoy, mover a mañana
    if (trigger.getTime() <= Date.now()) {
        trigger.setDate(trigger.getDate() + 1);
        // Re-establecer hora/minuto para no arrastrar el jitter al día siguiente
        trigger.setHours(slot.hour, slot.minute, 0, 0);
        if (jitterMinutes > 0) {
            const jitterMs = (Math.random() * 2 - 1) * jitterMinutes * 60 * 1000;
            trigger.setTime(trigger.getTime() + jitterMs);
        }
    }

    return trigger;
}

// ─────────────────────────────────────────────────
// Persistencia de franjas en @capacitor/preferences
// ─────────────────────────────────────────────────

/**
 * Lee las franjas guardadas por el usuario en Preferences.
 * Si no hay ninguna guardada (primera ejecución), devuelve DEFAULT_SLOTS.
 */
export async function getNotificationSlots(): Promise<TimeSlot[]> {
    try {
        const { value } = await Preferences.get({ key: PREFS_KEY });
        if (value) {
            const parsed = JSON.parse(value);
            if (Array.isArray(parsed) && parsed.length > 0) {
                return parsed as TimeSlot[];
            }
        }
    } catch (e) {
        console.warn('[NotificationSlots] Failed to read slots from preferences:', e);
    }
    return DEFAULT_SLOTS;
}

/**
 * Persiste las franjas elegidas por el usuario en Preferences.
 * Debe llamarse cuando el usuario guarda sus preferencias de horario.
 */
export async function saveNotificationSlots(slots: TimeSlot[]): Promise<void> {
    try {
        await Preferences.set({ key: PREFS_KEY, value: JSON.stringify(slots) });
    } catch (e) {
        console.warn('[NotificationSlots] Failed to save slots to preferences:', e);
    }
}
